lib/rex/ui/text/shell.rb

```diff
@@ -381,33 +381,46 @@ def prompt_yesno(query)
   #
   def format_prompt(str)
     if framework
-      if str.include?("%T")
+      if str.include?('%T')
         t = Time.now
         # This %T is the strftime shorthand for %H:%M:%S
-        format = framework.datastore['PromptTimeFormat'] || "%T"
+        format = framework.datastore['PromptTimeFormat'] || '%T'
         t = t.strftime(format)
         # This %T is the marker in the prompt where we need to place the time
-        str.gsub!(/%T/, t.to_s)
+        str.gsub!('%T', t.to_s)
       end
 
-      if str.include?("%H")
+      if str.include?('%H')
         hostname = ENV['HOSTNAME'] || `hostname`.split('.')[0] ||
           ENV['COMPUTERNAME'] || 'unknown'
 
-        str.gsub!(/%H/, hostname.chomp)
+        str.gsub!('%H', hostname.chomp)
       end
 
-      if str.include?("%U")
+      if str.include?('%U')
         user = ENV['USER'] || `whoami` || ENV['USERNAME'] || 'unknown'
-        str.gsub!(/%U/, user.chomp)
+        str.gsub!('%U', user.chomp)
       end
 
-      str.gsub!(/%S/, framework.sessions.length.to_s)
-      str.gsub!(/%J/, framework.jobs.length.to_s)
-      str.gsub!(/%L/, Rex::Socket.source_address("50.50.50.50"))
-      str.gsub!(/%D/, ::Dir.getwd)
-      if framework.db.active
-        str.gsub!(/%W/, framework.db.workspace.name)
+      if str.include?('%S')
+        str.gsub!('%S', framework.sessions.length.to_s)
+      end
+
+      if str.include?('%J')
+        str.gsub!('%J', framework.jobs.length.to_s)
+      end
+
+      if str.include?('%L')
+        ip = ::Socket.ip_address_list.detect{|intf| intf.ipv4_private?}
```

Comment:
Though `Rex::Socket.source_address` *shouldn't* send traffic to the destination, it still creates a UDP socket to my knowledge.

The behavior of `Rex::Socket.source_address` is different from `Socket.ip_address_list`'s, though, since it uses the socket to determine the correct IP that can talk to the destination. The behavior here is more like `ReverseListenerBindAddress` instead of `LHOST`.

```
[1] pry(#<Msf::Framework>)> Rex::Socket.source_address("50.50.50.50")
=> "66.x.x.x"
[2] pry(#<Msf::Framework>)> ::Socket.ip_address_list.detect{|intf| intf.ipv4_private?}
=> #<Addrinfo: 10.x.x.x>
[3] pry(#<Msf::Framework>)>
```

We may want to evaluate where else we're continuously leaking sockets, since `Rex::Socket.source_address` sees heavy use in Framework.

Reply:
If the problem is _leaking_ sockets the `source_address` code needs to be modified to put the socket close in an `ensure` block. Otherwise, this does have a big behavioral change in that it now grabs a local address instead of an internet address. https://github.com/rapid7/rex-socket/blob/master/lib/rex/socket.rb#L598-L605

Reply:
Exactly my thoughts. The `close` should be in an `ensure`. The code is already set up for it, even.

Reply:
I can just add a method for 'local_routable_ip' to Rex Socket as a class method and change Metasploit to use it everywhere? It seems you want to do something like this:

def self.local_routable_ip
    orig, Socket.do_not_reverse_lookup = Socket.do_not_reverse_lookup, true
    UDPSocket.open do |s|
      s.connect '50.50.50.50', 1
      s.addr.last
    end
    ensure
      Socket.do_not_reverse_lookup = orig
 end

Reply:
NM, I can just revert this to use rex and update that.